### PR TITLE
Ledger API W.R.T to customer and company

### PR DESCRIPTION
**Pull request by rehanyahya**

### CODE REVIEW

File: models/ledger.js [hunk 1]
```diff
@@ -2,23 +2,27 @@
...
   class ledger extends Model {
-    static associate({ Company, User }) {
+    static associate({ Company, Customer }) {
       this.belongsTo(Company, { foreignKey: "companyId" });
-      this.belongsTo(User, { foreignKey: "userId" });
+      this.belongsTo(Customer, { foreignKey: "customerId" });
     }
     toJSON() {
-      return { ...this.get(), companyId: undefined, userId: undefined };
+      return { ...this.get(), companyId: undefined };
```
Comment (DaniyalKukda) [inline models/ledger.js:10]: also add customerId undefined

File: pages/api/ledger/index.js [hunk 2]
```diff
@@ -20,12 +21,12 @@ const createTransaction = async (req, res) => {
     await db.dbConnect();
     const { user } = res;
 
-    const { totalAmount, companyId, spendType } = value;
+    const { totalAmount, companyId, spendType, customerId } = value;
```
Comment (DaniyalKukda) [inline pages/api/ledger/index.js:24]: what happen when customerId and companyId both are null?

File: pages/api/ledger/index.js [hunk 3]
```diff
@@ -20,12 +21,12 @@ const createTransaction = async (req, res) => {
     await db.dbConnect();
     const { user } = res;
```
Comment (DaniyalKukda) [inline pages/api/ledger/index.js:22]: get user from req not res

File: pages/api/ledger/index.js [hunk 4]
```diff
@@ -38,29 +39,47 @@ const getAllTransactions = async (req, res) => {
...
+    GROUP BY c."id"`;
 
-    return res.send({ transactions, ...balance });
+    const transactions = await db.sequelize.query(rawQuery, {
+      type: db.Sequelize.QueryTypes.SELECT,
+    });
+
+    // // const balance = transactions.reduce((a, b) => ({ totalBalance: a.total + b.total }));
+    let totalBalance = 0;
+    transactions.map((item) => (totalBalance += item.total));
```
Comment (DaniyalKukda) [inline pages/api/ledger/index.js:80]: i think we need to use reduce here instead of map

File: pages/api/ledger/index.js [hunk 5]
```diff
@@ -38,29 +39,47 @@ const getAllTransactions = async (req, res) => {
...
-      {
-        type: db.Sequelize.QueryTypes.SELECT,
-      }
-    );
+    const { type = "company" } = req.query;
 
-    const balance = transactions.reduce((a, b) => ({ totalBalance: a.total + b.total }));
+    const rawQuery =
+      type === "company"
+        ? `SELECT "companies"."companyName" as name,
```
Comment (DaniyalKukda) [inline pages/api/ledger/index.js:46]: we need to create new folder query and move all the raw queries there

File: models/ledger.js [hunk 6]
```diff
@@ -2,23 +2,27 @@
...
+        type: DataTypes.STRING,
+        allowNull: true,
+      },
       companyId: {
         type: DataTypes.INTEGER,
-        allowNull: false,
+        allowNull: true,
       },
-      userId: {
+      customerId: {
```
Comment (DaniyalKukda) [inline models/ledger.js:23]: if customerId and companyId is allow null Sequelize or postgres is allowed to null foreign keys columns?

File: pages/api/ledger/index.js (hunk 2 above)
Comment (rehanyahya) [reply pages/api/ledger/index.js:24]: It is not possible to set null both.

File: pages/api/ledger/index.js (hunk 3 above)
Comment (rehanyahya) [reply pages/api/ledger/index.js:22]: removed. not using it.

File: pages/api/ledger/index.js (hunk 2 above)
Comment (DaniyalKukda) [reply pages/api/ledger/index.js:22]: yes,

File: pages/api/ledger/index.js (hunk 4 above)
Comment (DaniyalKukda) [reply pages/api/ledger/index.js:78]: You need to use like this
`arr.reduce((acc , value) => acc + value  , 0)`

File: models/ledger.js (hunk 6 above)
Comment (DaniyalKukda) [reply models/ledger.js:23]: ok

File: pages/api/customer/index.js [hunk 7]
```diff
@@ -62,7 +67,7 @@ const getAllCustomers = async (req, res) => {
     });
   } catch (error) {
     console.log("Get all Customer Request Error:", error);
-    res.send(error);
+    res.send(500).send(error);
```
Comment (hammadtariq) [inline pages/api/customer/index.js:70]:
```suggestion
    res.status(500).send(error);
```

Comment (rehanyahya) [reply pages/api/customer/index.js:70]: done

**Pull request merged**